### DISCUSSION TITLE
Context should not use async

### DISCUSSION THREAD

**Discussion by BrynCooke** [apollographql/router#465] [Implementation]
Currently Context contains an async datastructure.

This has the following downsides:
* Eq/PartialEq cannot be derived.
* map_request and map_response can't store or retrieve in context as they are non async.
* Scripting plugins/Deno/WASM can't store or retrieve in context as they are non async.

Possible alternatives:
* Merge strategy - All contexts are merged on the way back and if there are non-unique results then they are discarded.
* ~~Synchronous lock - Should not have much contention, maybe this is OK?~~ Doesn't work.

**Comment by o0Ignition0o** [comment]
What’s our merge strategy ?
Prevent async / locking mechanisms

Maybe we shouldn’t provide Arc<Mutex<>> but require Send + Sync

Maybe adding convenience wrappers would suffice for now
